pkg/operator/certrotationcontroller/satokensigner_controller.go

```diff
@@ -91,6 +98,45 @@ func (c SATokenSignerController) sync() error {
 }
 
 func (c SATokenSignerController) syncWorker() error {
+	// we cannot rotate before the bootstrap server goes away because doing so would mean the bootstrap server would reject
+	// tokens that should be valid.  To test this, we go through kubernetes.default.svc endpoints and see if any of them
+	// are not in the list of known pod hosts.  We only have to do this once because the bootstrap node never comes back
+	if !c.confirmedBootstrapNodeGone {
+		nodeIPs := sets.String{}
+		apiServerPods, err := c.podClient.Pods("openshift-kube-apiserver").List(metav1.ListOptions{LabelSelector: "app=openshift-kube-apiserver"})
+		if err != nil {
+			return err
+		}
+		for _, pod := range apiServerPods.Items {
+			nodeIPs.Insert(pod.Status.HostIP)
+		}
+
+		kubeEndpoints, err := c.endpointClient.Endpoints("default").Get("kubernetes", metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if len(kubeEndpoints.Subsets) == 0 {
+			message := "missing kubernetes endpoints subsets"
+			c.eventRecorder.Warning("SATokenSignerControllerStuck", message)
+			return fmt.Errorf(message)
+		}
+		unexpectedEndpoints := sets.String{}
+		for _, subset := range kubeEndpoints.Subsets {
+			for _, address := range subset.Addresses {
+				if !nodeIPs.Has(address.IP) {
+					unexpectedEndpoints.Insert(address.IP)
+				}
+			}
+		}
+		if len(unexpectedEndpoints) != 0 {
+			message := fmt.Sprintf("unexpected addresses:", unexpectedEndpoints.List())
+			c.eventRecorder.Warning("SATokenSignerControllerStuck", message)
```

Comment:
is this really a warning? Should be normal during bootstrapping

Reply:
fixed